src/9FC031D0.c

```diff
@@ -1,33 +1,234 @@
 #include "include_asm.h"
 #include "ultratypes.h"
+#include "bbtypes.h"
+#include "libcrypto/sha1.h"
+#include "PR/bcp.h"
 
-INCLUDE_ASM("asm/non_matchings/9FC031D0", virage2_gen_public_key);
+void* wordcopy(void* dst, void* src, s32 nWords);
+void* memcpy(void* dst, void* src, u32 num);
 
-INCLUDE_ASM("asm/non_matchings/9FC031D0", set_proc_permissions);
+extern Virage2* virage2_offset;
 
-INCLUDE_ASM("asm/non_matchings/9FC031D0", rsa_verify_signature);
+void virage2_gen_public_key(u32* pubkeyOut) {
+    u32 i;
 
-INCLUDE_ASM("asm/non_matchings/9FC031D0", rsa_check_signature);
+    for(i = 0; i < 0x10; i++) {
```

Comment:
```suggestion
    for(i = 0; i < ARRAY_COUNT(virage2_offset->publicKey); i++) {
```
And same for the `wordcopy` calls below?

Reply:
Happy to do this, though I think ARRAY_COUNT isn't in yet (added in your currently open PR), so I didn't want to step on any toes and I figured we could clean it up later. I can just wait until yours is merged to address this.